### PR TITLE
feature/inline-typegen

### PR DESCRIPTION
**Pull request by bengsparks**
First initial implementation of inline type hint generation.
Extend the implementation during the evaluation period.

### CODE REVIEW

File: tests/typegen/strats/test_inline.py [hunk 1]
```diff
@@ -0,0 +1,378 @@
...
+
+    for node in ast.walk(hinted):
+        HintTest().visit(node)
+
+
+def test_assignments():
+    resource_path = pathlib.Path("tests", "resource", "typegen", "assignments.py")
+    assert resource_path.is_file()
+
+    gen = Generator(ident=InlineGenerator.ident, types=pd.DataFrame())
```
Comment (RecurvedBow) [inline tests/typegen/strats/test_inline.py:240]: Change to test_object and use the constructor of InlineGenerator directly.

File: tests/typegen/strats/test_inline.py [hunk 2]
```diff
@@ -0,0 +1,378 @@
...
+        str(resource_path),
+        c_clazz,
+        "method",
+        0,
+        TraceDataCategory.FUNCTION_RETURN,
+        "method",
+        "bytes",
+    ]
+
+    gen = Generator(ident=InlineGenerator.ident, types=traced)
```
Comment (RecurvedBow) [inline tests/typegen/strats/test_inline.py:224]: Same here

File: tracing/tracer.py [hunk 3]
```diff
@@ -164,30 +164,35 @@ def _on_trace_is_called(self, frame, event, arg: typing.Any) -> typing.Callable:
...
 
+            # Special case
+            line_number = 0
+
             # Adds tracing data of class members if the return is from a class function.
             if possible_class is not None:
                 names2types2 = self._on_class_function_return(frame)
                 category2 = TraceDataCategory.CLASS_MEMBER
-                self._update_trace_data_with(file_name, possible_class, "", 0, category2, names2types2)
+                self._update_trace_data_with(file_name, possible_class, "", line_number, category2, names2types2)
```
Comment (RecurvedBow) [inline tracing/tracer.py:186]: The former case with 0 is preferred as the line number variable name does not indicate that it is always 0.

File: typegen/strats/gen.py [hunk 4]
```diff
@@ -0,0 +1,64 @@
+import abc, ast
+import pathlib
+import typing
+
+import pandas as pd
+
+import constants
+
+
+class Generator(abc.ABC):
```
Comment (RecurvedBow) [inline typegen/strats/gen.py:10]: Choose a more precise name, for example, TypeHintGenerator, TypegenFileGenerator, ...

File: typegen/strats/inline.py [hunk 5]
```diff
@@ -0,0 +1,220 @@
...
+from tracing.trace_data_category import TraceDataCategory
+from typegen.strats.gen import Generator
+
+import pandas as pd
+
+logger = logging.getLogger(__name__)
+
+
+# It is important to handle sub-nodes from their parent node, as line numbering may differ across multi line statements
+class TypeHintApplierVisitor(ast.NodeTransformer):
```
Comment (RecurvedBow) [inline typegen/strats/inline.py:18]: Use -Transformer instead of -Visitor.

File: typegen/strats/gen.py [hunk 6]
```diff
@@ -0,0 +1,64 @@
...
+            applicable = self.types[
+                self.types[constants.TraceData.FILENAME] == str(path)
+            ]
+            if not applicable.empty:
+                nodes = ast.parse(source=path.open().read(), type_comments=True)
+                typed = self._gen_hints(df=applicable, nodes=nodes)
+                self._store_hints(source_file=path, hinting=typed)
+
+    @abc.abstractmethod
+    def _gen_hints(self, applicable: pd.DataFrame, nodes: ast.AST) -> ast.AST:
```
Comment (RecurvedBow) [inline typegen/strats/gen.py:53]: I recommend changing it to _gen_ast_with_hints.

File: typegen/strats/gen.py [hunk 7]
```diff
@@ -0,0 +1,64 @@
...
+        """
+        Perform operations to generate types for the given file
+        """
+        pass
+
+    @abc.abstractmethod
+    def _store_hints(self, source_file: pathlib.Path, hinting: ast.AST) -> None:
+        """
+        Perform operations to generate types for the given file
```
Comment (RecurvedBow) [inline typegen/strats/gen.py:62]: Same description as _gen_hints, unless it is intended.

File: typegen/strats/gen.py [hunk 8]
```diff
@@ -0,0 +1,64 @@
...
+
+    @abc.abstractmethod
+    def _gen_hints(self, applicable: pd.DataFrame, nodes: ast.AST) -> ast.AST:
+        """
+        Perform operations to generate types for the given file
+        """
+        pass
+
+    @abc.abstractmethod
+    def _store_hints(self, source_file: pathlib.Path, hinting: ast.AST) -> None:
```
Comment (RecurvedBow) [inline typegen/strats/gen.py:60]: The function name does not describe its functionality precise enough. Do you perhaps mean _on_ast_with_hints_generated?

**Review by RecurvedBow — CHANGES_REQUESTED** [review]
Looks pretty good. Mostly only small issues about naming.

File: tests/typegen/strats/test_inline.py (hunk 1 above)
Comment (bengsparks) [reply tests/typegen/strats/test_inline.py:240]: there is no constructor for InlineGenerator, and there needn't be one.
The factory method __new__ takes care of the set members

File: typegen/strats/gen.py (hunk 4 above)
Comment (bengsparks) [reply typegen/strats/gen.py:10]: `TypeHintGenerator` is good

File: typegen/strats/gen.py (hunk 6 above)
Comment (bengsparks) [reply typegen/strats/gen.py:53]: `_gen_hinted_ast`

File: typegen/strats/gen.py (hunk 7 above)
Comment (bengsparks) [reply typegen/strats/gen.py:62]: `_store_hinted_ast`

File: typegen/strats/inline.py (hunk 5 above)
Comment (bengsparks) [reply typegen/strats/inline.py]: Done

File: typegen/strats/gen.py (hunk 8 above)
Comment (bengsparks) [reply typegen/strats/gen.py]: `_store_hinted_ast`

File: tracing/tracer.py [hunk 9]
```diff
@@ -164,16 +164,21 @@ def _on_trace_is_called(self, frame, event, arg: typing.Any) -> typing.Callable:
...
             category = TraceDataCategory.FUNCTION_ARGUMENT
 
         elif event == "return":
-            logger.info(f"Tracing return: {inspect.getframeinfo(frame)}")
+            logger.info(f"Tracing return: {frameinfo}")
             names2types = self._on_return(frame, arg)
             category = TraceDataCategory.FUNCTION_RETURN
 
+            # Special case
+            line_number = 0
```
Comment (RecurvedBow) [inline tracing/tracer.py:180]: Remove these lines.

File: tracing/tracer.py (hunk 3 above)
Comment (bengsparks) [reply tracing/tracer.py]: Fixed

File: typegen/strats/inline.py [hunk 10]
```diff
@@ -0,0 +1,230 @@
...
+
+            arg_hint = arg_hints[TraceData.VARTYPE].values[0]
+            logger.debug(f"Applying hint '{arg_hint}' to '{arg.arg}'")
+            arg.annotation = ast.Name(arg_hint)
+
+        # disambiguate methods from functions
+        if hasattr(fdef, "parent"):
+            rettype_masks = [
+                self.df[TraceData.CATEGORY] == TraceDataCategory.FUNCTION_RETURN,
+                self.df[TraceData.CLASS] == fdef.parent.name,  # type: ignore
```
Comment (RecurvedBow) [inline typegen/strats/inline.py:74]: Is type: ignore necessary?

Comment (RecurvedBow) [reply typegen/strats/inline.py:74]: Also, will probably fail if the function is a nested function in a function. Perhaps you might want to check out line 124 - 130 to get the class name.

File: typegen/strats/inline.py [hunk 11]
```diff
@@ -0,0 +1,230 @@
...
+                self.df[TraceData.VARNAME] == fdef.name,
+                self.df[TraceData.LINENO] == 0,  # return type, always stored at line 0
+            ]
+
+        else:
+            rettype_masks = [
```
Comment (RecurvedBow) [inline typegen/strats/inline.py:80]: Check whether a class exists and merge the masks.

File: typegen/strats/inline.py [hunk 12]
```diff
@@ -0,0 +1,230 @@
...
+                self.df[TraceData.CATEGORY] == TraceDataCategory.FUNCTION_RETURN,
+                self.df[TraceData.CLASS].isnull(),
+                self.df[TraceData.VARNAME] == fdef.name,
+                self.df[TraceData.LINENO] == 0,
+            ]
+
+        rettypes = self.df[functools.reduce(operator.and_, rettype_masks)]
```
Comment (RecurvedBow) [inline typegen/strats/inline.py:87]: Consider naming it to function_return_trace_data.

File: typegen/strats/inline.py [hunk 13]
```diff
@@ -0,0 +1,230 @@
...
+        # NOTE: disregards *args (fdef.vararg) and **kwargs (fdef.kwarg)
+
+        # parameters
+        param_masks = [
+            self.df[TraceData.CATEGORY] == TraceDataCategory.FUNCTION_ARGUMENT,
+            self.df[TraceData.FUNCNAME] == fdef.name,
+        ]
+        params = self.df[functools.reduce(operator.and_, param_masks)]
+
+        for arg in itertools.chain(fdef.args.posonlyargs, fdef.args.args):
```
Comment (RecurvedBow) [inline typegen/strats/inline.py:50]: Put the handling of arguments and returns in different functions for better readability. for example: def add_type_hints_to_function_arguments & add_type_hints_to_function_return.

**Review by RecurvedBow — CHANGES_REQUESTED** [review]

File: tracing/tracer.py (hunk 9 above)
Comment (bengsparks) [reply tracing/tracer.py:180]: No, they are needed in the later call to update the data frame. it overwrites the number taken from the frame

File: typegen/strats/inline.py (hunk 10 above)
Comment (bengsparks) [reply typegen/strats/inline.py]: Added, fixes not finding return types on inner functions in methods

File: typegen/strats/inline.py (hunk 11 above)
Comment (bengsparks) [reply typegen/strats/inline.py]: Done

File: typegen/strats/inline.py (hunk 13 above)
Comment (bengsparks) [reply typegen/strats/inline.py:59]: Done

File: typegen/strats/inline.py [hunk 14]
```diff
@@ -0,0 +1,257 @@
...
+            names = list()
+            for target in targets:
+                names += self._extract_target_names_with_nodes(target)
+            return names
+
+    def visit_AugAssign(self, node: ast.AugAssign) -> ast.AST | list[ast.AST]:
+        assigns = self._visit_assigns(node)
+        # Overwrite with actual assignment
+        assert isinstance(assigns, list)
+        assigns[-1] = node
```
Comment (RecurvedBow) [inline typegen/strats/inline.py:132]: Should already be covered in line 218. Remove this line.

**Review by RecurvedBow — CHANGES_REQUESTED** [review]

File: typegen/strats/inline.py (hunk 14 above)
Comment (bengsparks) [reply typegen/strats/inline.py]: Done

File: typegen/strats/inline.py [hunk 15]
```diff
@@ -0,0 +1,262 @@
...
+        ]
+
+        local_traced_vars = self.df[
+            functools.reduce(operator.and_, var_mask_local_variables)
+        ]
+        class_members = self.df[functools.reduce(operator.and_, var_mask_class_members)]
+
+        # Attach hint directly to assignment and promote to AnnAssign
+        new_nodes: list[ast.AST] = []
+        contains_one_target = len(target_names_with_nodes) == 1
```
Comment (RecurvedBow) [inline typegen/strats/inline.py:185]: Add "and not isinstance(node, ast.AugAssign)" and change the name to only_update_node.

File: typegen/strats/inline.py [hunk 16]
```diff
@@ -0,0 +1,262 @@
...
+                    local_traced_vars[TraceData.VARNAME] == target_name
+                ]
+            if target_trace_data is None or len(target_trace_data) == 0:
+                logger.debug(f"No hint found for assign for '{target_name}'")
+                continue
+
+            logger.debug(f"Applying type hints for simple assignment '{target_name}'")
+            ann = target_trace_data[TraceData.VARTYPE].values[0].__name__
+
+            if contains_one_target and not isinstance(node, ast.AugAssign):
```
Comment (RecurvedBow) [inline typegen/strats/inline.py:207]: Update this.

**Review by RecurvedBow — CHANGES_REQUESTED** [review]

File: typegen/strats/inline.py (hunk 15 above)
Comment (bengsparks) [reply typegen/strats/inline.py]: Done

File: typegen/strats/inline.py (hunk 16 above)
Comment (bengsparks) [reply typegen/strats/inline.py]: Done

**Review by RecurvedBow — APPROVED** [review]

**Pull request merged**